File: matlab/code/3-signals-operations.m
clc;
clear;
close all;

% Original signal
n = -5:5;

x1 = (n >= -2 & n <= 2);
x2 = (n >= 0 & n <= 4);

% Addition
[x_add, n_add] = signalAddition(x1, n, x2, n);

% Folding
[x_fold, n_fold] = signalFolding(x1, n);

% Shifting
shift_value = 2;
[x_shift, n_shift] = signalShift(x1, n, shift_value);

% Plot Original Signals
figure;

subplot(2,2,1);
stem(n, x1, 'filled');
grid on;
xlabel('n');
ylabel('x_1[n]');
title('Signal x_1[n]');

subplot(2,2,2);
stem(n, x2, 'filled');
grid on;
xlabel('n');
ylabel('x_2[n]');
title('Signal x_2[n]');

% Addition
subplot(2,2,3);
stem(n_add, x_add, 'filled');
grid on;
xlabel('n');
ylabel('x_1[n] + x_2[n]');
title('Addition');

% Folding
figure;
stem(n_fold, x_fold, 'filled');
grid on;
xlabel('n');
ylabel('x_1[-n]');
title('Folding of x_1[n]');

% Shifting
figure;
stem(n_shift, x_shift, 'filled');
grid on;
xlabel('n');
ylabel('x_1[n-2]');
title('Right Shift by 2');

% ---------------- FUNCTIONS ----------------

function [y, n] = signalAddition(x1, n1, x2, n2)

    n = min([n1 n2]):max([n1 n2]);

    y1 = zeros(size(n));
    y2 = zeros(size(n));

    [~, index1] = ismember(n1, n);
    [~, index2] = ismember(n2, n);

    y1(index1) = x1;
    y2(index2) = x2;

    y = y1 + y2;
end

function [y, n_new] = signalFolding(x, n)

    y = fliplr(x);
    n_new = -fliplr(n);
end

function [y, n_new] = signalShift(x, n, k)

    n_new = n + k;
    y = x;
end

File: python/code/3-signals-operations.py
import numpy as np
import matplotlib.pyplot as plt

# Original signal
n = np.arange(-5, 6)

x = np.array([1, 2, 3, 4, 5, 4, 3, 2, 1, 0, 0])

# Second signal
h = np.array([0, 1, 2, 3, 4, 3, 2, 1, 0, 0, 0])


# -------------------------
# Addition
# -------------------------
addition = x + h


# -------------------------
# Folding
# x[-n]
# -------------------------
folded = x[::-1]


# -------------------------
# Shifting
# x[n-2]
# -------------------------
shift = 2

shifted_n = n + shift
shifted = x


# -------------------------
# Plot
# -------------------------
plt.figure(figsize=(12, 10))

plt.subplot(4, 1, 1)
plt.stem(n, x)
plt.title("Original Signal x[n]")
plt.xlabel("n")
plt.ylabel("x[n]")
plt.grid(True)

plt.subplot(4, 1, 2)
plt.stem(n, addition)
plt.title("Addition: x[n] + h[n]")
plt.xlabel("n")
plt.ylabel("Amplitude")
plt.grid(True)

plt.subplot(4, 1, 3)
plt.stem(n, folded)
plt.title("Folding: x[-n]")
plt.xlabel("n")
plt.ylabel("Amplitude")
plt.grid(True)

plt.subplot(4, 1, 4)
plt.stem(shifted_n, shifted)
plt.title("Shifting: x[n-2]")
plt.xlabel("n")
plt.ylabel("Amplitude")
plt.grid(True)

plt.tight_layout()
plt.show()